relay-server/src/services/buffer/envelope_stack/caching.rs

```diff
@@ -0,0 +1,101 @@
+use chrono::{DateTime, Utc};
+
+use super::EnvelopeStack;
+use crate::envelope::Envelope;
+
+/// An envelope stack implementation that caches one element in memory and delegates
+/// to another envelope stack for additional storage.
+#[derive(Debug)]
+pub struct CachingEnvelopeStack<S> {
+    /// The underlying envelope stack
+    inner: S,
+    /// The cached envelope (if any)
+    cached: Option<Box<Envelope>>,
+}
+
+impl<S> CachingEnvelopeStack<S>
+where
+    S: EnvelopeStack,
+{
+    /// Creates a new [`CachingEnvelopeStack`] wrapping the provided envelope stack
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner,
+            cached: None,
+        }
+    }
+}
+
+impl<S> EnvelopeStack for CachingEnvelopeStack<S>
+where
+    S: EnvelopeStack,
+{
+    type Error = S::Error;
+
+    async fn push(&mut self, envelope: Box<Envelope>) -> Result<(), Self::Error> {
+        if self.cached.is_none() {
+            // If we don't have a cached envelope, store this one in cache
+            self.cached = Some(envelope);
+            Ok(())
+        } else {
+            // If we already have a cached envelope, push the current cached one
+            // to the inner stack and keep the new one in cache
+            let cached = self.cached.take().unwrap();
+            self.inner.push(cached).await?;
+            self.cached = Some(envelope);
+            Ok(())
+        }
+    }
+
+    async fn peek(&mut self) -> Result<Option<DateTime<Utc>>, Self::Error> {
+        if let Some(ref envelope) = self.cached {
+            Ok(Some(envelope.received_at()))
+        } else {
+            self.inner.peek().await
+        }
```

Comment:
Tricky question whether peek should pop from the inner. Probably not.

Reply:
Isn't the `peek` behavior supposed always to return a reference if there is at least one envelope? Your idea seems to break this invariant.